### PR TITLE
GDB-6148 Add GraphDB interactive workload implementation

### PR DESCRIPTION
**Pull request by DenitsaStoianova**
Add LDBC SNB Interactive Workload GraphDB implementation
Change ldbc snb driver version from 0.3.5 to 0.3.6-SNAPSHOT

### CODE REVIEW

File: graphdb/src/main/java/com/ldbc/impls/workloads/ldbc/snb/graphdb/GraphDBConnectionState.java [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+
+		if (properties.containsKey("user") && properties.containsKey("password")) {
+			String user = properties.get("user");
+			String password = properties.get("password");
+			graphDBHTTPRepository.setUsernameAndPassword(user, password);
+		}
+	}
+
+	public RepositoryConnection getConnection() {
+		if (repositoryConnection == null) {
```
Comment (avataar) [inline graphdb/src/main/java/com/ldbc/impls/workloads/ldbc/snb/graphdb/GraphDBConnectionState.java]: No need for field repositoryConnection, you can simply always return graphDBHTTRepository.getConnection(). You can move the TimeZone thing to the constructor.

File: graphdb/src/main/java/com/ldbc/impls/workloads/ldbc/snb/graphdb/operationhandlers/GraphDBListOperationHandler.java [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+	public void executeOperation(TOperation operation, GraphDBConnectionState dbConnectionState, ResultReporter resultReporter)
+			throws DbException {
+		List<TOperationResult> results = new ArrayList<>();
+		int resultCount = 0;
+
+		final String queryString = getQueryString(dbConnectionState, operation);
+		try (RepositoryConnection conn = dbConnectionState.getConnection()) {
+			dbConnectionState.logQuery(operation.getClass().getSimpleName(), queryString);
+
+			TupleQueryResult queryResultIter = conn.prepareTupleQuery(QueryLanguage.SPARQL, queryString).evaluate();
```
Comment (avataar) [inline graphdb/src/main/java/com/ldbc/impls/workloads/ldbc/snb/graphdb/operationhandlers/GraphDBListOperationHandler.java]: TupleQueryResult needs to be closed - wrap it into try { }.

File: graphdb/src/main/java/com/ldbc/impls/workloads/ldbc/snb/graphdb/operationhandlers/GraphDBMultipleUpdateOperationHandler.java [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+
+import java.util.List;
+
+public abstract class GraphDBMultipleUpdateOperationHandler<TOperation extends Operation<LdbcNoResult>>
+		implements MultipleUpdateOperationHandler<TOperation, GraphDBConnectionState> {
+	@Override
+	public void executeOperation(TOperation operation, GraphDBConnectionState dbConnectionState, ResultReporter resultReporter) throws DbException {
+		List<String> queryStrings = getQueryString(dbConnectionState, operation);
+		try (RepositoryConnection conn = dbConnectionState.getConnection()) {
+			for (String queryString : queryStrings) {
```
Comment (avataar) [inline graphdb/src/main/java/com/ldbc/impls/workloads/ldbc/snb/graphdb/operationhandlers/GraphDBMultipleUpdateOperationHandler.java:22]: When executing multiple updates it's best to start a transaction before all updates and commit the transaction after all the updates.

File: graphdb/src/main/java/com/ldbc/impls/workloads/ldbc/snb/graphdb/operationhandlers/GraphDBSingletonOperationHandler.java [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+	public void executeOperation(TOperation operation, GraphDBConnectionState dbConnectionState, ResultReporter resultReporter)
+			throws DbException {
+		TOperationResult tuple = null;
+		int resultCount = 0;
+
+		final String queryString = getQueryString(dbConnectionState, operation);
+		try (RepositoryConnection conn = dbConnectionState.getConnection()) {
+			dbConnectionState.logQuery(operation.getClass().getSimpleName(), queryString);
+
+			TupleQueryResult queryResult = conn.prepareTupleQuery(QueryLanguage.SPARQL, queryString).evaluate();
```
Comment (avataar) [inline graphdb/src/main/java/com/ldbc/impls/workloads/ldbc/snb/graphdb/operationhandlers/GraphDBSingletonOperationHandler.java]: TupleQueryResult needs to be closed - wrap it into try { }.

File: graphdb/src/main/java/com/ldbc/impls/workloads/ldbc/snb/graphdb/converter/GraphDBConverter.java [hunk 5]
```diff
@@ -0,0 +1,119 @@
...
+	//converts date of type "2011-08-12" to millisecond.
+	public long localDateToEpoch(BindingSet bindingSet, String name) {
+		String s = bindingSet.getValue(name).stringValue();
+		return LocalDate.parse(s).atStartOfDay(ZoneId.of("GMT")).toInstant().toEpochMilli();
+	}
+
+	public Iterable<String> asStringCollection(BindingSet bindingSet, String name) {
+		String stringValue = bindingSet.getValue(name).stringValue();
+		if (stringValue.isEmpty()) {
+			return new ArrayList<>();
```
Comment (avataar) [inline graphdb/src/main/java/com/ldbc/impls/workloads/ldbc/snb/graphdb/converter/GraphDBConverter.java]: If you need an empty list use Collections.emptyList()

File: graphdb/src/main/java/com/ldbc/impls/workloads/ldbc/snb/graphdb/converter/GraphDBConverter.java [hunk 6]
```diff
@@ -0,0 +1,119 @@
...
+		} else {
+			return Arrays.asList(stringValue.split(COLLECTION_SEPARATOR));
+		}
+	}
+
+	public Iterable<Number> asNumberCollection(BindingSet bindingSet, String name) {
+		String stringValue = bindingSet.getValue(name).stringValue();
+		if (stringValue.isEmpty()) {
+			return new ArrayList<>();
```
Comment (avataar) [inline graphdb/src/main/java/com/ldbc/impls/workloads/ldbc/snb/graphdb/converter/GraphDBConverter.java]: If you need an empty list use Collections.emptyList()

File: graphdb/src/main/java/com/ldbc/impls/workloads/ldbc/snb/graphdb/converter/GraphDBConverter.java [hunk 7]
```diff
@@ -0,0 +1,119 @@
...
+		} else {
+			List<String> stringList = Arrays.asList(stringValue.split(COLLECTION_SEPARATOR));
+			return stringList.stream().map(Long::parseLong).collect(Collectors.toList());
+		}
+	}
+
+	public Iterable<List<Object>> asObjectCollection(BindingSet bindingSet, String name) {
+		String stringValue = bindingSet.getValue(name).stringValue();
+		if (stringValue.isEmpty()) {
+			return new ArrayList<>();
```
Comment (avataar) [inline graphdb/src/main/java/com/ldbc/impls/workloads/ldbc/snb/graphdb/converter/GraphDBConverter.java]: If you need an empty list use Collections.emptyList()

**Review by avataar — DISMISSED** [review]

File: graphdb/config/docker-compose-start.yml [hunk 8]
```diff
@@ -0,0 +1,24 @@
+version: "3"
+
+services:
+  graphdb:
+    container_name: ${GRAPHDB_CONTAINER_NAME}
+    image: ontotext/graphdb:${GRAPHDB_VERSION}-free
```
Comment (nikolaykolev) [inline graphdb/config/docker-compose-start.yml]: There's no such image in the docker hub. You should update the readme to include info how to build an image with the free version and set it here

**Review by tokovach — APPROVED** [review]

**Pull request merged**